lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -1549,6 +1549,19 @@ class DestructureInputs {
       auto eltPattern = origType.getFunctionParamType(i);
       auto flags = params[i].getParameterFlags();
 
+      // If we have an inline operator, when it was imported, it was turned
+      // from a method into a static function. So, we need to shift over the
+      // params by one and use "self" as the first param's type.
```

Comment:
I can't say what the right fix should be exactly, but I feel like this code is working around the problem. Resetting `NextOrigParamIndex` does not feel right to me. Abstraction pattern and the code to handle 'self' in this converter class are supposed to abstract away this issue, I think. Currently they handle the abstraction in the opposite direction (importing a C free function as a Swift method). Here, we have the opposite problem -- importing a C++ member function as a Swift static method.

WDYT about extending this code to skip both the Swift `self` which was mapped from C, and Swift parameters that correspond to C++ `this`, and letting the `HandleForeignSelf` handle them at the appropriate point?

Reply:
> WDYT about extending this code to skip both the Swift self which was mapped from C, and Swift parameters that correspond to C++ this, and letting the HandleForeignSelf handle them at the appropriate point?

The issue is `CXXMethodConventions` creates a `CFunctionTypeConventions` object with the function type `T(T)` when it should be `T(T, T)`. This means that there are the wrong number of arguments. I think we have two choices, either we can pass `forSelf=true` for the first argument, then decrement `NextOrigParamIndex` (still a bit of a hack but also less bad than what's currently in this PR), or we can pass a more "correct" type to `CFunctionTypeConventions`. WDYT would be better?

Reply:
I've updated this PR to use the first solution I suggested but I can change it to use the second if you'd rather.

Reply:
@hlopko @varungandhi-apple thoughts on the above conversation? I'd like to get this patch moving :)

Reply:
@gribozavr @hlopko ping. Once re-reviewed I'll rebase.